arangod/Aql/AstNode.cpp

```diff
@@ -376,6 +383,57 @@ static bool IsEmptyString(char const* p, size_t length) {
   return true;
 }
 
+/*static*/ std::string AstNode::encodeDataSourceType(
+    char* const name, size_t size, AstNode::DataSourceType type
+) {
+  StringRef suffix;
+
+  switch (type) {
+    case AstNode::DataSourceType::View:
+      suffix = VIEW_NODE_SUFFIX;
+      break;
+    case AstNode::DataSourceType::Collection:
+      suffix = COLLECTION_NODE_SUFFIX;
+      break;
+    default: // shut up compiler
+      break;
+  }
+
+  std::string param(name, size + suffix.size());
+  std::memcpy(&param[0] + size, suffix.data(), suffix.size());
+  return param;
+}
+
+/*static*/ AstNode::DataSourceType AstNode::decodeDataSourceType(
+    std::string& param
+) {
+  auto decodeParam = [](std::string& param, StringRef const& expectedSuffix) -> bool {
+    auto const suffixPos = param.size() - expectedSuffix.size();
+
+    StringRef const suffix(
+      param.c_str() + suffixPos,
+      expectedSuffix.size()
+    );
+
+    if (suffix == expectedSuffix) {
+      param.resize(suffixPos);
```

Comment:
isn't param.size() always the same as suffixPos when we get here?

Reply:
ok, it's only equal when called from Ast::injectBindParameters, not from grammar.y

Reply:
no, consider the following string: `testView\0v`
param.size() == 10
suffixPos == 8

Reply:
yes, but when it is called from Ast::injectBindParameters, I think params.size() is always the same as suffixPos.